nixos/modules/programs/wayland/uwsm.nix

```diff
@@ -0,0 +1,117 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+let
+  cfg = config.programs.uwsm;
+
+  # Helper function to create desktop entry files for UWSM-managed compositors
+  mk_uwsm_desktop_entry =
+    opts:
+    (pkgs.writeTextFile {
+      name = "${opts.name}_uwsm";
+      text = ''
+        [Desktop Entry]
+        Name=${opts.compositorPrettyName} (UWSM)
+        Comment=${opts.compositorComment}
+        Exec=${lib.getExe cfg.package} start -S -F "${opts.compositorBinPath}"
+        Type=Application
+      '';
+      destination = "/share/wayland-sessions/${opts.name}-uwsm.desktop";
+      derivationArgs = {
+        passthru.providedSessions = [ "${opts.name}-uwsm" ];
+      };
+    });
+in
+{
+  options.programs.uwsm = {
+    enable = lib.mkEnableOption ''
+      uwsm, which wraps standalone Wayland compositors with a set
+      of Systemd units on the fly. This essentially
+      binds the wayland compositor into `graphical-session-pre.target`,
+      `graphical-session.target`, `xdg-desktop-autostart.target`.
+
+      This is useful for Wayland compositors like Hyprland, Sway, Wayfire,
+      etc. that do not start these targets and services on their own.
+
+      If you are having trouble starting a service that depends on
+      `graphical-session.target`, while using a WM, enabling and
+      configuring this might help
+    '';
+    package = lib.mkPackageOption pkgs "uwsm" { };
+
+    waylandCompositors = lib.mkOption {
+      description = ''
+        Configuration for UWSM-managed Wayland Compositors.
+        This creates a desktop entry file which will be used by
+        Display Managers to allow starting the UWSM managed session.
+      '';
+      type = lib.types.attrsOf (
+        lib.types.submodule (
+          { ... }:
+          {
+            options = {
+              # name,
+              # Name, Comment (for desktop entry)
+              # Compositor-Package -> The main binary could be extracted from getExe
+              # which would then be used in `Exec` of desktop entry
+              compositorPrettyName = lib.mkOption {
+                type = lib.types.str;
+                description = "The full name of the desktop entry file.";
+                example = "ExampleWaylandCompositor";
```

Comment:
I personally would favor a "real" example, like Hyprland/Sway just to improve clarity.
Though I do not have a strong opinion here.

Reply:
I included that just above.

https://github.com/NixOS/nixpkgs/blob/1cf2e9be457d75a71f6abfbc30dbf4b17219c5a4/nixos/modules/programs/wayland/uwsm.nix#L81-L92